File: java-web-spring-boot-3/src/main/java/com/vndevteam/javawebspringboot3/infrastructure/exception/ExceptionsHandler.java
package com.vndevteam.javawebspringboot3.infrastructure.exception;

import com.vndevteam.javawebspringboot3.infrastructure.exception.error.ErrorCode;
import com.vndevteam.javawebspringboot3.infrastructure.exception.error.ErrorDetail;
import com.vndevteam.javawebspringboot3.infrastructure.exception.error.ErrorResponseDto;
import com.vndevteam.javawebspringboot3.util.DateTimeUtils;
import jakarta.validation.ConstraintViolationException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.ConversionNotSupportedException;
import org.springframework.beans.TypeMismatchException;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.http.converter.HttpMessageNotWritableException;
import org.springframework.lang.NonNull;
import org.springframework.util.CollectionUtils;
import org.springframework.validation.method.MethodValidationException;
import org.springframework.web.ErrorResponseException;
import org.springframework.web.HttpMediaTypeNotAcceptableException;
import org.springframework.web.HttpMediaTypeNotSupportedException;
import org.springframework.web.HttpRequestMethodNotSupportedException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.MissingPathVariableException;
import org.springframework.web.bind.MissingServletRequestParameterException;
import org.springframework.web.bind.ServletRequestBindingException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.context.request.async.AsyncRequestTimeoutException;
import org.springframework.web.method.annotation.HandlerMethodValidationException;
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException;
import org.springframework.web.multipart.MaxUploadSizeExceededException;
import org.springframework.web.multipart.support.MissingServletRequestPartException;
import org.springframework.web.servlet.NoHandlerFoundException;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
import org.springframework.web.servlet.resource.NoResourceFoundException;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.ArrayList;

@RestControllerAdvice
@Slf4j
public class ExceptionsHandler extends ResponseEntityExceptionHandler {
    @Value("${app.logging.debug}")
    private boolean debug;

    @Override
    protected ResponseEntity<Object> handleHttpRequestMethodNotSupported(
            HttpRequestMethodNotSupportedException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        log.warn(ex.getMessage());

        var errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.REQUEST_METHOD_INVALID,
                ex.getMessage(),
                null,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.BAD_REQUEST);
    }

    @Override
    protected ResponseEntity<Object> handleHttpMediaTypeNotSupported(
            HttpMediaTypeNotSupportedException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        log.warn(ex.getMessage());

        var errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.MEDIA_TYPE_INVALID,
                ex.getMessage(),
                null,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.BAD_REQUEST);
    }

    @Override
    protected ResponseEntity<Object> handleHttpMediaTypeNotAcceptable(
            @NonNull HttpMediaTypeNotAcceptableException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleHttpMediaTypeNotAcceptable(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleMissingPathVariable(
            @NonNull MissingPathVariableException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleMissingPathVariable(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleMissingServletRequestParameter(
            MissingServletRequestParameterException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        ErrorResponseDto errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.VALIDATION_INVALID,
                ex.getMessage(),
                null,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.BAD_REQUEST);
    }

    @Override
    protected ResponseEntity<Object> handleMissingServletRequestPart(
            @NonNull MissingServletRequestPartException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleMissingServletRequestPart(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleServletRequestBindingException(
            ServletRequestBindingException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        var errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.REQUEST_HEADER_INVALID,
                ex.getMessage(),
                null,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.BAD_REQUEST);
    }

    @Override
    protected ResponseEntity<Object> handleMethodArgumentNotValid(
            MethodArgumentNotValidException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        var errors = new ArrayList<ErrorDetail>();
        if (CollectionUtils.isEmpty(ex.getBindingResult().getFieldErrors())) {
            for (var error : ex.getBindingResult().getAllErrors()) {
                errors.add(new ErrorDetail(null, null, error.getDefaultMessage()));
            }
        } else {
            for (var error : ex.getBindingResult().getFieldErrors()) {
                errors.add(new ErrorDetail(null, error.getField(), error.getDefaultMessage()));
            }
        }

        log.warn(errors.toString());

        var errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.VALIDATION_INVALID,
                "Method argument not valid",
                errors,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.BAD_REQUEST);
    }

    @Override
    protected ResponseEntity<Object> handleHandlerMethodValidationException(
            @NonNull HandlerMethodValidationException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleHandlerMethodValidationException(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleNoHandlerFoundException(
            @NonNull NoHandlerFoundException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleNoHandlerFoundException(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleNoResourceFoundException(
            @NonNull NoResourceFoundException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleNoResourceFoundException(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleAsyncRequestTimeoutException(
            @NonNull AsyncRequestTimeoutException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleAsyncRequestTimeoutException(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleErrorResponseException(
            @NonNull ErrorResponseException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleErrorResponseException(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleMaxUploadSizeExceededException(
            @NonNull MaxUploadSizeExceededException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleMaxUploadSizeExceededException(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleConversionNotSupported(
            @NonNull ConversionNotSupportedException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleConversionNotSupported(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleTypeMismatch(
            @NonNull TypeMismatchException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleTypeMismatch(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleHttpMessageNotReadable(
            @NonNull HttpMessageNotReadableException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleHttpMessageNotReadable(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleHttpMessageNotWritable(
            @NonNull HttpMessageNotWritableException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode status,
            @NonNull WebRequest request) {
        return super.handleHttpMessageNotWritable(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleMethodValidationException(
            @NonNull MethodValidationException ex,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatus status,
            @NonNull WebRequest request) {
        return super.handleMethodValidationException(ex, headers, status, request);
    }

    @Override
    protected ResponseEntity<Object> handleExceptionInternal(
            @NonNull Exception ex,
            Object body,
            @NonNull HttpHeaders headers,
            @NonNull HttpStatusCode statusCode,
            @NonNull WebRequest request) {
        return super.handleExceptionInternal(ex, body, headers, statusCode, request);
    }

    @ExceptionHandler(MethodArgumentTypeMismatchException.class)
    public ResponseEntity<ErrorResponseDto> handleMethodArgumentTypeMismatch(MethodArgumentTypeMismatchException ex) {
        var errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.VALIDATION_INVALID,
                ex.getMessage(),
                null,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<ErrorResponseDto> handleConstraintViolation(ConstraintViolationException ex) {
        var errors = new ArrayList<ErrorDetail>();
        for (var violation : ex.getConstraintViolations()) {
            errors.add(new ErrorDetail(null, null, violation.getMessage()));
        }

        log.warn(errors.toString());

        var errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.VALIDATION_INVALID,
                ex.getMessage(),
                null,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.BAD_REQUEST);
    }

    /**
     * Default handler Handle all system errors
     *
     * @param ex Exception
     * @return ResponseEntity<ErrorResponseDto>
     */
    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponseDto> handleOthers(Exception ex) {
        log.error("Error: {}", ex.getMessage());

        var errorDetail = new ErrorResponseDto(
                DateTimeUtils.getNow(),
                ErrorCode.SYSTEM_ERROR,
                "System error",
                null,
                getServerMessage(ex));

        return new ResponseEntity<>(errorDetail, HttpStatus.INTERNAL_SERVER_ERROR);
    }

    private String getServerMessage(Exception ex) {
        if (debug) {
            var sw = new StringWriter();
            ex.printStackTrace(new PrintWriter(sw));

            return sw.toString();
        }

        return null;
    }
}

File: kotlin-web-spring-boot-3/src/main/kotlin/com/vndevteam/kotlinwebspringboot3/infrastructure/exception/ExceptionsHandler.kt
package com.vndevteam.kotlinwebspringboot3.infrastructure.exception

import com.vndevteam.kotlinwebspringboot3.infrastructure.exception.error.ErrorDetail
import com.vndevteam.kotlinwebspringboot3.infrastructure.exception.error.ErrorResponseDto
import com.vndevteam.kotlinwebspringboot3.util.DateTimeUtils
import jakarta.validation.ConstraintViolationException
import java.io.PrintWriter
import java.io.StringWriter
import java.util.*
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.util.CollectionUtils
import org.springframework.web.HttpMediaTypeNotSupportedException
import org.springframework.web.HttpRequestMethodNotSupportedException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.MissingServletRequestParameterException
import org.springframework.web.bind.ServletRequestBindingException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import org.springframework.web.method.annotation.MethodArgumentTypeMismatchException
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler

@RestControllerAdvice
class ExceptionsHandler : ResponseEntityExceptionHandler() {
    companion object {
        val log: Logger = LoggerFactory.getLogger(ExceptionsHandler::class.java)
    }

    @Value("\${app.logging.debug}") val debug: Boolean = false

    /**
     * Handle when method argument is not the expected type
     *
     * @param ex
     * @return
     */
    @ExceptionHandler(MethodArgumentTypeMismatchException::class)
    fun handleMethodArgumentTypeMismatch(
        ex: MethodArgumentTypeMismatchException
    ): ResponseEntity<Any> {

        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.VALIDATION_INVALID,
                message = ex.message,
                trace = getServerMessage(ex),
                errors = null
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Handle when sending a request with an unsupported HTTP method
     *
     * @param ex
     * @param headers
     * @param status
     * @param request
     * @return
     */
    override fun handleHttpRequestMethodNotSupported(
        ex: HttpRequestMethodNotSupportedException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any>? {

        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.REQUEST_METHOD_INVALID,
                message = ex.message,
                trace = getServerMessage(ex),
                errors = null
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Handle when sending a request with unsupported media type
     *
     * @param ex
     * @param headers
     * @param status
     * @param request
     * @return
     */
    override fun handleHttpMediaTypeNotSupported(
        ex: HttpMediaTypeNotSupportedException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any>? {
        log.warn(ex.message)

        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.MEDIA_TYPE_INVALID,
                message = ex.message,
                trace = getServerMessage(ex),
                errors = null
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Handle all invalid attributes
     *
     * @param ex
     * @param headers
     * @param status
     * @param request
     * @return
     */
    override fun handleMethodArgumentNotValid(
        ex: MethodArgumentNotValidException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any>? {
        val errors: List<ErrorDetail> =
            if (CollectionUtils.isEmpty(ex.bindingResult.fieldErrors)) {
                ex.bindingResult.allErrors.map { e ->
                    ErrorDetail(code = null, field = null, message = e.defaultMessage)
                }
            } else {
                ex.bindingResult.fieldErrors.map { e ->
                    ErrorDetail(code = null, field = e.field, message = e.defaultMessage)
                }
            }
        log.warn(errors.toString())

        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.VALIDATION_INVALID,
                message = "Method argument not valid",
                trace = getServerMessage(ex),
                errors = errors
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Handle when parameter is invalid (using for spring @Validated annotation)
     *
     * @param ex
     * @return
     */
    @ExceptionHandler(ConstraintViolationException::class)
    fun handleConstraintViolation(ex: ConstraintViolationException): ResponseEntity<Any> {
        val errors =
            ex.constraintViolations.map { e ->
                ErrorDetail(code = null, field = null, message = e.message)
            }
        log.warn(errors.toString())

        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.VALIDATION_INVALID,
                message = ex.message,
                trace = getServerMessage(ex),
                errors = errors
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Handle when request missing parameter
     *
     * @param ex
     * @param headers
     * @param status
     * @param request
     * @return
     */
    override fun handleMissingServletRequestParameter(
        ex: MissingServletRequestParameterException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any> {
        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.VALIDATION_INVALID,
                message = ex.message,
                trace = getServerMessage(ex),
                errors = null
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Handle when sending a request with header invalid
     *
     * @param ex
     * @param headers
     * @param status
     * @param request
     * @return
     */
    override fun handleServletRequestBindingException(
        ex: ServletRequestBindingException,
        headers: HttpHeaders,
        status: HttpStatusCode,
        request: WebRequest
    ): ResponseEntity<Any>? {
        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.REQUEST_HEADER_INVALID,
                message = ex.message,
                trace = getServerMessage(ex),
                errors = null
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    /**
     * Default handler Handle all system errors
     *
     * @param ex
     * @return
     */
    @ExceptionHandler(Exception::class)
    fun handleOthers(ex: Exception): ResponseEntity<Any> {
        log.error(ex.message)

        return ResponseEntity(
            ErrorResponseDto(
                timestamp = DateTimeUtils.getNow(),
                code = ErrorConstants.SYSTEM_ERROR,
                message = "System error",
                trace = getServerMessage(ex),
                errors = null
            ),
            HttpStatus.INTERNAL_SERVER_ERROR
        )
    }

    private fun getServerMessage(ex: Exception): String? {
        if (debug) {
            val sw = StringWriter()
            ex.printStackTrace(PrintWriter(sw))

            return sw.toString()
        }

        return null
    }
}
